### ISSUE TITLE
'unhandled ZLE widget' errors when KSH_ARRAYS is set

### ISSUE THREAD

**Issue by okdana** [zsh-users/zsh-syntax-highlighting#162]
zsh has an option called `KSH_ARRAYS` which is very useful for portability and has the following effects:

1. Arrays are indexed from 0 instead of 1
2. Unqualified references to arrays in parameter expansion (`$arr`) return the first member instead of the entire array
3. References to array members in parameter expansion must include delimiting braces (`${arr[foo]}`, not `$arr[foo]`)

When this option is enabled, zsh-syntax-highlighting ceases to work, primarily because of difference # 3. Specifically, the lack of braces on `$widgets[$cur_widget]` in `_zsh_highlight_bind_widgets()` creates a screen full of errors like this (after which highlighting is disabled):

```
zsh-syntax-highlighting: unhandled ZLE widget 'accept-and-hold'
zsh-syntax-highlighting: unhandled ZLE widget 'accept-and-infer-next-history'
zsh-syntax-highlighting: unhandled ZLE widget 'accept-and-menu-complete'
zsh-syntax-highlighting: unhandled ZLE widget 'accept-line'
zsh-syntax-highlighting: unhandled ZLE widget 'accept-line-and-down-history'
zsh-syntax-highlighting: unhandled ZLE widget 'accept-search'
zsh-syntax-highlighting: unhandled ZLE widget 'argument-base'
zsh-syntax-highlighting: unhandled ZLE widget 'auto-suffix-remove'
zsh-syntax-highlighting: unhandled ZLE widget 'auto-suffix-retain'
...
```

Simply adding braces around the array references in these functions would fix 90% of the issue, but it looks like there are actually one or two places where difference # 1 would also come into play.

Therefore, to resolve all of these differences, i would suggest that functions dealing with arrays could have the following line added to the top:

```
setopt localoptions noksharrays
```

This will disable `KSH_ARRAYS` for the duration of the function call.

http://zsh.sourceforge.net/Guide/zshguide02.html

**Comment by danielshahaf** [comment]
Workaround: change

    source /path/to/zsh-syntax-highlighting/zsh-syntax-highlighting.zsh

to

    emulate -R zsh -c 'source /path/to/zsh-syntax-highlighting/zsh-syntax-highlighting.zsh'

in `.zshrc`.

**Comment by danielshahaf** [comment]
Should be fixed by 8f19af6.

**Comment by danielshahaf** [comment]
For future reference, we can't simply `emulate -L` in the driver, since the individual highlighters need to know some pre-emulate option values: https://github.com/zsh-users/zsh-syntax-highlighting/blob/8f19af6b319db6130f376f3f2a27b5b74acd0562/highlighters/main/main-highlighter.zsh#L85-L95

It's not impossible, it just means we have to save the `$options` array (or the output of `$(setopt kshoptionprint; setopt)` if the `zsh/parameter` module is unavailable) somewhere available to the highlighters.

**Comment by danielshahaf** [comment]
Now that this specific problem has been fixed, issue #233 tracks preventing future problems of this kind.
